x/mm2/keeper/msg_server.go

```diff
@@ -0,0 +1,136 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/skip-mev/slinky/x/mm2/types"
+)
+
+// msgServer is the default implementation of the x/marketmap MsgService.
+type msgServer struct {
+	k *Keeper
+}
+
+// NewMsgServer returns the default implementation of the x/marketmap message service.
+func NewMsgServer(k *Keeper) types.MsgServer {
+	return &msgServer{k}
+}
+
+var _ types.MsgServer = (*msgServer)(nil)
+
+// CreateMarkets updates the marketmap by creating markets from the given message.  All updates are made to the market
+// map and then the resulting final state is checked to verify that the end state is valid.
+func (ms msgServer) CreateMarkets(goCtx context.Context, msg *types.MsgCreateMarkets) (*types.MsgCreateMarketsResponse, error) {
+	if msg == nil {
+		return nil, fmt.Errorf("unable to process nil msg")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	params, err := ms.k.GetParams(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get marketmap params: %w", err)
+	}
+
+	found := checkMarketAuthority(msg.Authority, params)
+	if !found {
+		return nil, fmt.Errorf("request signer %s does not match module market authorities", msg.Authority)
+	}
+
+	// create markets
+	for _, createMarket := range msg.CreateMarkets {
+		err = ms.k.CreateMarket(ctx, createMarket)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO hooks
+
+		// TODO events
+	}
+
+	// validate that the new state of the marketmap is valid
+	err = ms.k.ValidateState(ctx, msg.CreateMarkets)
+	if err != nil {
+		return nil, fmt.Errorf("invalid state resulting from update: %w", err)
+	}
```

Comment:
do we need to delete the newly created markets?

Reply:
No - any msg server level function in cosmos sdk is a "transaction", so its "all or nothing" like a db transaction.

The state is only applied if the entire tx is valid